Mathlib/Algebra/Operad/Basic.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2024 Alex Meiburg. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Meiburg
+-/
+import Mathlib.Algebra.Group.Action.Defs
+
+/-! This file defines the typeclasses and notations used for workign with `Operad`, `SymmOperad`,
+  and `Clone`. These all, in some form another, work with graded - or Sigma - types. Below,
+  `A : ℕ → Type*` is always the carrier.
+
+  Classes:
+  - `MultiComposable`: A `compose` operation `A n → Fin n → A m → A (n+m-1)`, for composing an
+    element of grade `m` into one of grade `n` at a location `Fin n`. This is the type of operation
+    used in Operads.
+  - `Superposable`: A `superpose` operation `A n → (Fin n → A m) → A m`. This is the type of
+    operation used in Clones.
+  - `OneGradedOne`: There is a distinguished `1` element at grade 1, a notion of identity.
+  - `SigmaMulAction`: A graded `MulAction M A`, where elements `M i` can act on `A i` at each i.
+
+  Definitions:
+  - `composeAt` bundles the action of `MultiComposable.compose` into `Sigma A`.
+  - `superpose` bundles the action of `Superposable.superpose` into `Sigma A`.
+
+  Notations:
+  - `x ∘⟨p⟩ y` for `MultiComposable.compose` to compose `x ∘ y` at index `p`.
+  - `x ∘[p] y` for the `composeAt` Sigma variant.
+  - `x ∘⚟ y` for `Superposable.superpose x y`. The typography is meant to suggest
+    "many arguments into one".
+-/
+
+/-- A MultiComposable is a structure that allows composition from an m-arity object
+ into a n-arity object at location p (in the range 0 to n-1) to produce an (n+m-1)
+ arity object. `Operad` is the canonical example. -/
+class MultiComposable  (A : ℕ → Type*) where
+  /-- Compose the (n+1)-arity object at location p on the m-arity object. -/
+  compose {n m} : A n → Fin n → A m → A (n+m-1)
```

Comment:
What is `n+m-1` supposed to be when `n = m = 0`?
Depending on the answer, you may consider the following design:
```lean
  compose {n m x} : A n → Fin n → A m → n + m = x + 1 → A x
```

Reply:
As `n` must be nonzero, this cannot happen... However, my suggestion of using `n + m = x + 1` still stands.

Reply:
I think this would be less typical. But I added it as an alternative, `composeEq`, that unfolds to this, with a helper theorem.